src/Spectre.libHeatmapDataScaling/HistogramEqualization.cpp

```diff
@@ -0,0 +1,73 @@
+/*
+* HistogramEqualization.cpp
+* Class with histogram equalization algorithm implementation for contrast enhancement.
+*
+Copyright 2018 Daniel Babiak
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#include "Spectre.libHeatmapDataScaling/HistogramEqualization.h"
+
+namespace spectre::visualization
+{
+std::vector<double> HistogramEqualization::scaleData(const gsl::span<double> intensities)
+{
+    std::vector<double> roundedIntensities = roundIntensities(intensities);
+    std::map<double, unsigned int> const histogramMap = calculateCumulativeDistribution(countRepeatingValues(roundedIntensities));
+    return calculateNewHistogramData(roundedIntensities, histogramMap);
+}
+
+std::vector<double> HistogramEqualization::roundIntensities(const gsl::span<double> intensities)
+{
+    std::vector<double> roundedIntensities(intensities.size());
+    for (int i = 0; i < intensities.size(); i++)
+    {
+        roundedIntensities[i] = round(intensities[i]);
+    }
+    return roundedIntensities;
+}
+
+std::map<double, unsigned int> HistogramEqualization::countRepeatingValues(const gsl::span<double> intensities)
+{
+    std::map<double, unsigned int> histogramMap;
+    for (auto intensity : intensities)
+    {
+        histogramMap[intensity]++;
```

Comment:
We did not scale the values to fixed range before counting and rounding. Is this intended? This will work differently for the same input data just scaled by a constant factor (rounding will have proportionally different impact on the data).

BTW. I don't think we will ever need a `map`, if the data is scaled to 0-255 at the beginning. Starting from such point we may just operate on `vector` with 256 elements.